packages/markers/src/browser/marker-tree.ts

```diff
@@ -0,0 +1,140 @@
+*
+* Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+*/
+
+import { injectable, inject } from "inversify";
+import { Tree, ICompositeTreeNode, ITreeNode, ISelectableTreeNode, IExpandableTreeNode } from "@theia/core/lib/browser";
+import { MarkerManager, Marker } from './marker-manager';
+import { UriSelection } from "@theia/filesystem/lib/common";
+import URI from "@theia/core/lib/common/uri";
+import { JSONExt, ReadonlyJSONObject } from "@phosphor/coreutils";
+
+export const MarkerOptions = Symbol('MarkerOptions');
+export interface MarkerOptions {
+    readonly kind: string;
+}
+
+@injectable()
+export class MarkerTree extends Tree {
+
+    constructor(
+        @inject(MarkerManager) protected readonly markerManager: MarkerManager,
+        @inject(MarkerOptions) protected readonly markerOptions: MarkerOptions
+    ) {
+        super();
+
+        markerManager.onDidChangeMarkers(() => this.refresh());
+
+        this.root = <MarkerRootNode>{
+            visible: false,
+            id: 'theia-' + markerOptions.kind + '-marker-widget',
+            name: 'MarkerTree',
+            kind: markerOptions.kind,
+            children: [],
+            parent: undefined
+        };
+    }
+
+    resolveChildren(parent: ICompositeTreeNode): Promise<ITreeNode[]> {
+        if (MarkerRootNode.is(parent)) {
+            return this.getMarkerInfoNodes((parent as MarkerRootNode));
+        } else if (MarkerInfoNode.is(parent)) {
+            return this.getMarkerNodes(parent);
+        }
+        return super.resolveChildren(parent);
+    }
+
+    getMarkerInfoNodes(parent: MarkerRootNode): Promise<MarkerInfoNode[]> {
+        const uriNodes: MarkerInfoNode[] = [];
+        if (this.root && MarkerRootNode.is(this.root)) {
+            this.markerManager.forEachMarkerInfoByKind(this.root.kind, markerInfo => {
+                const uri = new URI(markerInfo.uri);
+                const id = 'markerInfo-' + markerInfo.uri;
+                const cachedMarkerInfo = this.getNode(id);
+                if (cachedMarkerInfo && MarkerInfoNode.is(cachedMarkerInfo)) {
+                    cachedMarkerInfo.numberOfMarkers = markerInfo.counter;
+                    uriNodes.push(cachedMarkerInfo);
+                } else {
+                    uriNodes.push({
+                        children: [],
+                        expanded: false,
+                        uri,
+                        id,
+                        name: uri.displayName,
+                        parent,
+                        selected: false,
+                        numberOfMarkers: markerInfo.counter
+                    });
+                }
+            });
+        }
+        return Promise.resolve(uriNodes);
+    }
+
+    getNodeByMarker(marker: Marker<object>): MarkerNode | undefined {
```

Comment:
I think it can be overly expensive, and we better move the marker data comparison to `MarkerCollection.setMarkers` to reuse ids from existing markers, the search scope will be narrowed to one owner and a URI instead of looping over all tree nodes for each new marker